module/Solr/src/Solr/Listener/JobEventSubscriber.php

```diff
@@ -49,154 +73,101 @@ public function __construct(Manager $manager)
     public function getSubscribedEvents()
     {
         return [
+            Events::preUpdate,
             Events::postUpdate,
-            Events::postPersist,
+            Events::postFlush
         ];
     }
-    public function consoleIndex(Job $job)
-    {
-        $this->updateIndex($job);
-    }
+    
     /**
-     * Handle doctrine post persist event
-     *
      * @param LifecycleEventArgs $eventArgs
+     * @since 0.27
      */
-    public function postPersist(LifecycleEventArgs $eventArgs)
+    public function preUpdate(PreUpdateEventArgs $eventArgs)
     {
         $document = $eventArgs->getDocument();
-        $this->updateIndex($document);
+        
+        // check for a job instance
+        if (!$document instanceof Job) {
+            return;
+        }
+        
+        // check if the status has been changed
+        if (!$eventArgs->hasChangedField('status')) {
+            return;
+        }
+            
+        // check if the job is active
+        if ($document->isActive()) {
+            // mark it for commit
+            $this->add[] = $document;
+        } else {
+            // mark it for delete
+            $this->delete[] = $document;
+        }
     }
+    
     /**
-     * Handle doctrine postUpdate event
-     *
      * @param LifecycleEventArgs $eventArgs
      */
     public function postUpdate(LifecycleEventArgs $eventArgs)
     {
-        $document = $eventArgs->getDocument();
-        $this->updateIndex($document);
-    }
-    /**
-     * @param ServiceLocatorInterface $serviceLocator
-     * @return mixed
-     */
-    static public function factory(ServiceLocatorInterface $serviceLocator)
-    {
-        /* @var Manager $manager */
-        $manager = $serviceLocator->get('Solr/Manager');
-        return new self($manager);
-    }
-    /**
-     * @param $document
-     */
-    protected function updateIndex($document)
-    {
-        if(!$document instanceof Job){
+        // check if there is any job to process
+        if (!$this->add && !$this->delete) {
             return;
         }
-        $solrDoc = $this->generateInputDocument($document, new \SolrInputDocument());
-        try{
-            $this->solrManager->addDocument($solrDoc,$this->solrManager->getOptions()->getJobsPath());
-        }catch (\Exception $e){
-            // @TODO: What to do when the process failed?
+        
+        $client = $this->getSolrClient();
+        
+        // process jobs for commit
+        foreach ($this->add as $job) {
+            $document = $this->entityToDocumentFilter->filter($job);
+            $client->addDocument($document);
+        }
+        
+        // process jobs for delete
+        foreach ($this->delete as $job) {
+            $client->deleteByIds($this->entityToDocumentFilter->getDocumentIds($job));
         }
     }
+    
```

Comment:
Isn't postUpdate hook called after EACH document?

Maybe we do not need this hook and handle all of it in the postFlush hook?